File: sdk/js/test/crc8.test.ts
import { describe, it, expect } from 'vitest'
import { crc8 } from '../src/core/crc8.js'

describe('CRC8 Dallas/Maxim', () => {
  it('returns 0 for empty input', () => {
    expect(crc8(new Uint8Array([]))).toBe(0x00)
  })

  it('computes CRC for single byte 0x00', () => {
    expect(crc8(new Uint8Array([0x00]))).toBe(0x00)
  })

  it('computes CRC for single byte 0x01', () => {
    expect(crc8(new Uint8Array([0x01]))).toBe(0x31)
  })

  it('computes CRC for known vector [0x01, 0x02]', () => {
    const result = crc8(new Uint8Array([0x01, 0x02]))
    expect(result).toBeTypeOf('number')
    expect(result).toBeGreaterThanOrEqual(0)
    expect(result).toBeLessThanOrEqual(255)
  })

  it('produces different CRC for different data', () => {
    const a = crc8(new Uint8Array([0x01, 0x02, 0x03]))
    const b = crc8(new Uint8Array([0x01, 0x02, 0x04]))
    expect(a).not.toBe(b)
  })

  it('is deterministic', () => {
    const data = new Uint8Array([0x47, 0x46, 0x01, 0x02, 0x00, 0x00, 0x00])
    const a = crc8(data)
    const b = crc8(data)
    expect(a).toBe(b)
  })

  it('computes CRC for a longer payload', () => {
    const data = new Uint8Array(256)
    for (let i = 0; i < 256; i++) data[i] = i
    const result = crc8(data)
    expect(result).toBeTypeOf('number')
    expect(result).toBeGreaterThanOrEqual(0)
    expect(result).toBeLessThanOrEqual(255)
  })
})

File: sdk/go/crc8_test.go
package graft

import "testing"

func TestCRC8Empty(t *testing.T) {
	if got := CRC8([]byte{}); got != 0x00 {
		t.Errorf("CRC8(empty) = 0x%02x, want 0x00", got)
	}
}

func TestCRC8SingleZero(t *testing.T) {
	if got := CRC8([]byte{0x00}); got != 0x00 {
		t.Errorf("CRC8([0x00]) = 0x%02x, want 0x00", got)
	}
}

func TestCRC8Single0x01(t *testing.T) {
	if got := CRC8([]byte{0x01}); got != 0x31 {
		t.Errorf("CRC8([0x01]) = 0x%02x, want 0x31", got)
	}
}

func TestCRC8Deterministic(t *testing.T) {
	data := []byte{0x47, 0x46, 0x01, 0x02}
	a := CRC8(data)
	b := CRC8(data)
	if a != b {
		t.Errorf("CRC8 not deterministic: %02x != %02x", a, b)
	}
}

func TestCRC8DifferentData(t *testing.T) {
	a := CRC8([]byte{0x01, 0x02, 0x03})
	b := CRC8([]byte{0x01, 0x02, 0x04})
	if a == b {
		t.Error("CRC8 should produce different values for different data")
	}
}
